src/signatures/commitment_signature.rs

```diff
@@ -85,88 +89,131 @@ where
+        let u_x = r_x + &ex;
+        let u_y = r_y + &ey;
 
-        Ok(Self::new(public_commitment_nonce, u, v))
+        // Compute the initial values
+        let ephemeral_commitment = factory.commit(r_x, r_a);
+        let ephemeral_pubkey = P::from_secret_key(r_y);
+
+        Ok(Self::new(ephemeral_commitment, ephemeral_pubkey, u_a, u_x, u_y))
     }
 
-    /// Verify if the commitment signature signed the commitment using the specified challenge (as bytes). If the
-    /// provided challenge n bytes cannot be converted to a secret key, this function also returns false.
+    /// Verify a signature on a commitment and group element statement using a given challenge (as a byte array)
     pub fn verify_challenge<'a, C>(
         &self,
-        public_commitment: &'a HomomorphicCommitment<P>,
+        commitment: &'a HomomorphicCommitment<P>,
+        pubkey: &'a P,
         challenge: &[u8],
         factory: &C,
     ) -> bool
     where
         for<'b> &'a HomomorphicCommitment<P>: Mul<&'b K, Output = HomomorphicCommitment<P>>,
+        for<'b> &'a P: Mul<&'b K, Output = P>,
         for<'b> &'b HomomorphicCommitment<P>: Add<&'b HomomorphicCommitment<P>, Output = HomomorphicCommitment<P>>,
+        for<'b> &'b P: Add<&'b P, Output = P>,
         C: HomomorphicCommitmentFactory<P = P>,
     {
+        // The challenge must be a valid scalar
         let e = match K::from_bytes(challenge) {
             Ok(e) => e,
             Err(_) => return false,
         };
 
-        self.verify(public_commitment, &e, factory)
+        self.verify(commitment, pubkey, &e, factory)
     }
 
-    /// Verify if the commitment signature signed the commitment using the specified challenge (as secret key).
-    ///  v*H + u*G = R + e.C
-    pub fn verify<'a, C>(&self, public_commitment: &'a HomomorphicCommitment<P>, challenge: &K, factory: &C) -> bool
+    /// Verify a signature on a commitment and group element statement using a given challenge (as a scalar)
+    pub fn verify<'a, C>(&self, commitment: &'a HomomorphicCommitment<P>, pubkey: &'a P, challenge: &K, factory: &C) -> bool
     where
         for<'b> &'a HomomorphicCommitment<P>: Mul<&'b K, Output = HomomorphicCommitment<P>>,
+        for<'b> &'a P: Mul<&'b K, Output = P>,
         for<'b> &'b HomomorphicCommitment<P>: Add<&'b HomomorphicCommitment<P>, Output = HomomorphicCommitment<P>>,
+        for<'b> &'b P: Add<&'b P, Output = P>,
         C: HomomorphicCommitmentFactory<P = P>,
     {
-        // v*H + u*G
-        let lhs = self.calc_signature_verifier(factory);
-        // R + e.C
-        let rhs = &self.public_nonce + &(public_commitment * challenge);
-        // Implementors should make this a constant time comparison
-        lhs == rhs
+        // The challenge cannot be zero
+        if challenge.to_owned() == K::default() {
+            return false
+        }
+
+        // TODO: Use efficient multiscalar multiplication with a random verifier weight between the equations
```

Comment:
Why is this `// TODO`? Is it not as straight forward as presented in the comment^^?

Reply:
It wasn't clear to me if it was straightforward or possible to use multiscalar multiplication evaluation given the generic key types, nor if it's even an efficiency win in the non-batched case.

Reply:
This is (partially) done in a more recent commit. Verification now merges the two equations into a single weighted equation, but still calls into the commitment factory to take advantage of future precomputation there. It would be most ideal to use precomputation with curve library multiscalar multiplication, but that can be left to future work if desired.